srtp/srtp.c

```diff
@@ -121,6 +127,182 @@ static size_t srtp_get_rtp_xtn_hdr_len(const srtp_hdr_t *hdr,
     return (ntohs(xtn_hdr->length) + 1u) * 4u;
 }
 
+static uint16_t srtp_get_rtp_xtn_hdr_profile(const srtp_hdr_t *hdr,
+                                             const uint8_t *rtp)
+{
+    const srtp_hdr_xtnd_t *xtn_hdr =
+        (const srtp_hdr_xtnd_t *)(rtp + srtp_get_rtp_hdr_len(hdr));
+    return ntohs(xtn_hdr->profile_specific);
+}
+
+static void srtp_cryptex_adjust_buffer(const srtp_hdr_t *hdr, uint8_t *rtp)
+{
+    uint8_t tmp[4];
+    uint8_t *ptr = rtp + srtp_get_rtp_hdr_len(hdr);
+    memcpy(tmp, ptr, 4);
+    for (size_t i = hdr->cc; i > 0; --i) {
+        memcpy(ptr, ptr - 4, 4);
+        ptr -= 4;
+    }
+    memcpy(ptr, tmp, 4);
+}
+
+static void srtp_cryptex_restore_buffer(const srtp_hdr_t *hdr, uint8_t *rtp)
+{
+    uint8_t tmp[4];
+    uint8_t *ptr = rtp + octets_in_rtp_header;
+    memcpy(tmp, ptr, 4);
+    for (size_t i = 0; i < hdr->cc; ++i) {
+        memcpy(ptr, ptr + 4, 4);
+        ptr += 4;
+    }
+    memcpy(ptr, tmp, 4);
+}
+
+static srtp_err_status_t srtp_cryptex_protect_init(
+    const srtp_stream_ctx_t *stream,
+    const srtp_hdr_t *hdr,
+    const uint8_t *rtp,
+    const uint8_t *srtp,
+    bool *inuse,
+    bool *inplace,
+    size_t *enc_start)
+{
+    if (stream->use_cryptex && (stream->rtp_services & sec_serv_conf)) {
+        if (hdr->cc && hdr->x == 0) {
+            /* Cryptex can only encrypt CSRCS if header extension is present */
+            return srtp_err_status_cryptex_err;
+        }
+        *inuse = hdr->x == 1;
+    } else {
+        *inuse = false;
+    }
+
+    *inplace = *inuse && rtp == srtp;
+
+    if (*inuse) {
+        *enc_start -=
+            (srtp_get_rtp_xtn_hdr_len(hdr, rtp) - octets_in_rtp_xtn_hdr);
+        if (*inplace) {
+            *enc_start -= (hdr->cc * 4);
+        }
+    }
+
+    return srtp_err_status_ok;
+}
+
+static srtp_err_status_t srtp_cryptex_protect(bool inplace,
+                                              const srtp_hdr_t *hdr,
+                                              uint8_t *srtp,
+                                              srtp_cipher_t *rtp_cipher)
+{
+    srtp_hdr_xtnd_t *xtn_hdr = srtp_get_rtp_xtn_hdr(hdr, srtp);
+    uint16_t profile = ntohs(xtn_hdr->profile_specific);
+    if (profile == xtn_hdr_one_byte_profile) {
+        xtn_hdr->profile_specific = htons(cryptex_one_byte_profile);
+    } else if (profile == xtn_hdr_two_byte_profile) {
+        xtn_hdr->profile_specific = htons(cryptex_two_byte_profile);
+    } else {
+        return srtp_err_status_parse_err;
+    }
+
+    if (inplace) {
+        srtp_cryptex_adjust_buffer(hdr, srtp);
+    } else {
+        if (hdr->cc) {
+            uint8_t *cc_list = srtp + octets_in_rtp_header;
+            size_t cc_list_size = hdr->cc * 4;
+            /* CSRCs are in dst header already, enc in place */
+            srtp_err_status_t status = srtp_cipher_encrypt(
```

Comment:
Maybe we should talk about this, but it's not clear to me how this works. The "inplace" version, which is used with AEAD, makes sense since it moves the CSRCs to form the plaintext as described in 6.2/RFC9335. However, this seems to encrypt the CC list separately. I'm confused.

Reply:
Talked with @pabuhler. I now understand what I was missing.